deepsubtree.go

```diff
@@ -0,0 +1,271 @@
+// using the given full IAVL tree along everything along
+// the path of that node
+func (dst *DeepSubTree) AddPath(tree *ImmutableTree, key []byte) error {
+	path, val, err := tree.root.PathToLeaf(tree, key)
+	if err != nil {
+		return err
+	}
+
+	err = dst.addPath(path, val)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Helper method to add given leaf node in the Deep Subtree
+// using the given PathToLeaf
+func (dst *DeepSubTree) addPath(pl PathToLeaf, leaf *Node) error {
+	hash, err := leaf._hash()
+	if err != nil {
+		return err
+	}
+
+	n := NewNode(leaf.key, leaf.value, leaf.version)
+	n._hash()
+	err = dst.ndb.SaveNode(n)
+	if err != nil {
+		return err
+	}
+
+	prevHash := n.hash
+	for i := len(pl) - 1; i >= 0; i-- {
+		pin := pl[i]
+		hash, err = pin.Hash(hash)
+		if err != nil {
+			return err
+		}
+		if pin.Left == nil {
+			pin.Left = prevHash
+		} else if pin.Right == nil {
+			pin.Right = prevHash
+		}
+		n := &Node{
+			subtreeHeight: pin.Height,
+			size:          pin.Size,
+			version:       pin.Version,
+			leftHash:      pin.Left,
+			rightHash:     pin.Right,
+			hash:          hash,
+		}
+		prevHash = hash
+
+		has, err := dst.ndb.Has(n.hash)
+		if err != nil {
+			return err
+		}
+		if !has {
+			err = dst.ndb.SaveNode(n)
+			if err != nil {
+				return err
+			}
+		}
+		if i == 0 {
+			if err := dst.ndb.Commit(); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Traverses in the nodes in the NodeDB in the Deep Subtree
+// and links them together using the populated left and right
+// hashes and sets the root to be the node with the given rootHash
+func (dst *DeepSubTree) BuildTree(rootHash []byte) error {
+	nodes, traverseErr := dst.ndb.nodes()
+	if traverseErr != nil {
+		return fmt.Errorf("could not traverse nodedb: %w", traverseErr)
+	}
+	// Traverse through nodes and link them correctly
+	for _, node := range nodes {
+		pnode, _ := dst.ndb.GetNode(node.hash)
+		if len(pnode.leftHash) > 0 && pnode.leftNode == nil {
+			pnode.leftNode, _ = dst.ndb.GetNode(pnode.leftHash)
+		}
+		if len(pnode.rightHash) > 0 && pnode.rightNode == nil {
+			pnode.rightNode, _ = dst.ndb.GetNode(pnode.rightHash)
+		}
+	}
+	// Now that nodes are linked correctly, traverse again
+	// and set their keys correctly
+	for _, node := range nodes {
+		pnode, _ := dst.ndb.GetNode(node.hash)
+		if pnode.leftNode != nil {
+			pnode.key = pnode.leftNode.getHighestKey()
+		}
+
+		if pnode.rightNode != nil {
+			pnode.key = pnode.rightNode.getLowestKey()
+		}
+	}
+	if dst.root == nil {
+		rootNode, rootErr := dst.ndb.GetNode(rootHash)
+		if rootErr != nil {
+			return fmt.Errorf("could not set root of deep subtree: %w", rootErr)
+		}
+		dst.root = rootNode
+	}
+
+	return nil
+}
+
+// Set sets a key in the working tree with the given value.
+// Assumption: Node with given key already exists and is a leaf node.
+func (dst *DeepSubTree) Set(key []byte, value []byte) (updated bool, err error) {
+	if value == nil {
```

Comment:
Would you change it in the `mutable_tree.go` implementation as well then? https://github.com/celestiaorg/iavl/blob/56e223198068a06c3fc1bad141b2444965ee5929/mutable_tree.go#L225

Reply:
Looked into `iavl` a little. So a `nil` key doesn't cause any error cases it seems, which is probably why it isn't explicitly checked for.  Since a `nil` key will result in an empty string as the key.

So my question would be should we allow blank keys?

This is definitely more of a follow up discussion that we could have in slack maybe.
Feels like a user error to submit a `nil` key that we would have to protect against, but maybe not. 

fine to resolve.